### PR TITLE
feat:adds set-enforcement-action function for policy constraints

### PR DESCRIPTION
**Pull request by g-awmalik**
This addresses [this issue](https://github.com/GoogleContainerTools/kpt/issues/2460).

### CODE REVIEW

File: examples/set-enforcement-action-simple/resources.yaml [hunk 1]
```diff
@@ -0,0 +1,30 @@
+apiVersion: constraints.gatekeeper.sh/v1beta1
+kind: K8sRestrictRoleBindings
+metadata: # kpt-merge: /restrict-clusteradmin-rolebindings
+  name: restrict-clusteradmin-rolebindings
+  annotations:
+    # This constraint is not certified by CIS.
+    description: "Restricts use of the cluster-admin role."
+    bundles.validator.forsetisecurity.org/cis-k8s-v1.5.1: 5.1.1
+spec:
+  enforcementAction: deny # kpt-set: ${enforcementAction}
```
[the remark this answers is not in the document]
Comment (g-awmalik) [reply examples/set-enforcement-action-simple/resources.yaml:10]: Agreed and acked.

File: examples/set-enforcement-action-simple/resources.yaml [hunk 2]
```diff
@@ -0,0 +1,30 @@
+apiVersion: constraints.gatekeeper.sh/v1beta1
+kind: K8sRestrictRoleBindings
```
[the remark this answers is not in the document]
Comment (g-awmalik) [reply examples/set-enforcement-action-simple/resources.yaml:2]: This fn isn't intended to validate a constraint locally like `policy-controller-validate` does. This is only flips the enforcement action element to dryrun, deny or warn (thanks for pointing that out) based on the config passed.

File: functions/go/set-enforcement-action/README.md [hunk 3]
```diff
@@ -0,0 +1,42 @@
...
+
+<!--mdtogo:Short-->
+
+Applies the supplied enforcement action on [policy constraints](https://cloud.google.com/anthos-config-management/docs/concepts/policy-controller#constraints) within a package.
+
+<!--mdtogo-->
+
+Policy Controller [constraint objects](https://cloud.google.com/anthos-config-management/docs/how-to/creating-constraints) enable you to enforce policies for your Kubernetes clusters.
+This function provides a quick way for users to set the [enforcementAction](https://cloud.google.com/anthos-config-management/docs/how-to/auditing-constraints#types_of_enforcement_actions)
+attribute to either `dryrun` for auditing or `deny` for enforcing the constraints.
```
[the remark this answers is not in the document]
Comment (g-awmalik) [reply functions/go/set-enforcement-action/README.md:13]: I wasn't aware that `warn` was added as well. I'll update the README for this.

File: functions/go/set-enforcement-action/main.go [hunk 4]
```diff
@@ -0,0 +1,132 @@
...
+	denyActionValue      = "deny"
+	dryRunActionValue    = "dryrun"
+)
+
+//nolint
+func main() {
+	fp := SetEnforcementActionProcessor{}
+	cmd := command.Build(&fp, command.StandaloneEnabled, false)
+	if err := cmd.Execute(); err != nil {
+		fmt.Fprintln(os.Stderr, err)
```
[the remark this answers is not in the document]
Comment (g-awmalik) [reply functions/go/set-enforcement-action/main.go:26]: Acked.

File: functions/go/set-enforcement-action/main.go [hunk 5]
```diff
@@ -0,0 +1,132 @@
...
+			continue
+		}
+
+		//check for resources of type policy
+		metadata, err := node.GetMeta()
+		if err != nil {
+			return nil, err
+		}
+
+		if metadata.Name == "" || metadata.Kind == "" || metadata.APIVersion != policyAPIVersion {
```
[the remark this answers is not in the document]
Comment (g-awmalik) [reply functions/go/set-enforcement-action/main.go:66]: Technically it may not be but all CIS policies have names set and I thought it's better for a policy to have a name instead of skipping it. It also helps us monitor the usage internally when we run our metrics.

File: functions/go/set-enforcement-action/main.go [hunk 6]
```diff
@@ -0,0 +1,132 @@
...
+			continue
+		}
+
+		acnElem, err := node.Pipe(yaml.Lookup("spec", "enforcementAction"))
+		if err != nil {
+			log.Fatal(err)
```
[the remark this answers is not in the document]
Comment (g-awmalik) [reply functions/go/set-enforcement-action/main.go:72]: My bad. Acked.

File: functions/go/set-enforcement-action/main.go [hunk 7]
```diff
@@ -0,0 +1,132 @@
...
+		metadata, err := node.GetMeta()
+		if err != nil {
+			return nil, err
+		}
+
+		if metadata.Name == "" || metadata.Kind == "" || metadata.APIVersion != policyAPIVersion {
+			continue
+		}
+
+		acnElem, err := node.Pipe(yaml.Lookup("spec", "enforcementAction"))
```
[the remark this answers is not in the document]
Comment (g-awmalik) [reply functions/go/set-enforcement-action/main.go:70]: Sure, that's better. Changed it accordingly.

File: functions/go/set-enforcement-action/main.go [hunk 8]
```diff
@@ -0,0 +1,132 @@
...
+		if err != nil {
+			log.Fatal(err)
+		}
+
+		acnElem.YNode().Value = acn
```
[the remark this answers is not in the document]
Comment (g-awmalik) [reply functions/go/set-enforcement-action/main.go:75]: Never mind, saw your next comment. I changed this accordingly.

File: functions/go/set-enforcement-action/main.go (hunk 7 above)
[the remark this answers is not in the document]
Comment (g-awmalik) [reply functions/go/set-enforcement-action/main.go:70]: Agreed, this is better.  I've changed it now.

File: functions/go/set-enforcement-action/main.go (hunk 8 above)
[the remark this answers is not in the document]
Comment (g-awmalik) [reply functions/go/set-enforcement-action/main.go:75]: Never mind. Saw your next comment and I've changed it accordingly.

**Comment by g-awmalik** [comment]
@mengqiy - thanks for the review. Please take another look.

File: functions/go/set-enforcement-action/README.md [hunk 9]
```diff
@@ -0,0 +1,42 @@
...
+## Overview
+
+<!--mdtogo:Short-->
+
+Applies the supplied enforcement action on [policy constraints](https://cloud.google.com/anthos-config-management/docs/concepts/policy-controller#constraints) within a package.
+
+<!--mdtogo-->
+
+Policy Controller [constraint objects](https://cloud.google.com/anthos-config-management/docs/how-to/creating-constraints) enable you to enforce policies for your Kubernetes clusters.
+This function provides a quick way for users to set the [enforcementAction](https://cloud.google.com/anthos-config-management/docs/how-to/auditing-constraints#types_of_enforcement_actions)
```
Comment (mengqiy) [inline functions/go/set-enforcement-action/README.md:12]: > [enforcementAction](...

If referring a field, please use a full path and backquote it. i.e. 
```
[`spec.enforcementAction`](...
```

File: functions/go/set-enforcement-action/README.md [hunk 10]
```diff
@@ -0,0 +1,42 @@
...
+
+## Overview
+
+<!--mdtogo:Short-->
+
+Applies the supplied enforcement action on [policy constraints](https://cloud.google.com/anthos-config-management/docs/concepts/policy-controller#constraints) within a package.
+
+<!--mdtogo-->
+
+Policy Controller [constraint objects](https://cloud.google.com/anthos-config-management/docs/how-to/creating-constraints) enable you to enforce policies for your Kubernetes clusters.
```
Comment (mengqiy) [inline functions/go/set-enforcement-action/README.md:11]: May want to reword it to:
`Policy Controller allows users to define [constraints](https://cloud.google.com/anthos-config-management/docs/how-to/creating-constraints) to enforce policies for your Kubernetes clusters.`

File: functions/go/set-enforcement-action/README.md [hunk 11]
```diff
@@ -0,0 +1,42 @@
...
+```
+
+To execute `set-enforcement-action` declaratively include the function in kpt package pipeline as follows:
+```yaml
+...
+pipeline:
+  mutators:
+    - image: gcr.io/kpt-fn/set-enforcement-action:unstable
+      configMap:
+          enforcementAction: deny
```
Comment (mengqiy) [inline functions/go/set-enforcement-action/README.md:38]: Indentation is not consistent with other lines (2 whitespaces).

File: functions/go/set-enforcement-action/README.md [hunk 12]
```diff
@@ -0,0 +1,42 @@
...
+
+<!--mdtogo:Short-->
+
+Applies the supplied enforcement action on [policy constraints](https://cloud.google.com/anthos-config-management/docs/concepts/policy-controller#constraints) within a package.
+
+<!--mdtogo-->
+
+Policy Controller [constraint objects](https://cloud.google.com/anthos-config-management/docs/how-to/creating-constraints) enable you to enforce policies for your Kubernetes clusters.
+This function provides a quick way for users to set the [enforcementAction](https://cloud.google.com/anthos-config-management/docs/how-to/auditing-constraints#types_of_enforcement_actions)
+attribute to either `dryrun` for auditing, `warn` for letting the non-compliant resource be applied to the cluster with a warning or `deny` for enforcing the constraints and denying the resource application altogether.
```
Comment (mengqiy) [inline functions/go/set-enforcement-action/README.md:13]: We should not use `either` since we have 3 options here. Please fix the wording

File: functions/go/set-enforcement-action/main.go [hunk 13]
```diff
@@ -0,0 +1,131 @@
...
+}
+
+// getEnforcementAction gets the value to set for enforcementAction from the functionConfig
+func getEnforcementAction(fc *yaml.RNode, acn *string) error {
+	if len(fc.GetDataMap()) != 1 {
+		return errors.New("expecting exactly 1 enforcementAction as part of the ConfigMap")
+	}
+
+	*acn = fc.GetDataMap()[enforcementActionKey]
+	if *acn != denyActionValue && *acn != dryRunActionValue {
```
Comment (mengqiy) [inline functions/go/set-enforcement-action/main.go:114]: Not support `warn` here.

File: examples/set-enforcement-action-simple/README.md [hunk 14]
```diff
@@ -0,0 +1,64 @@
...
+apiVersion: constraints.gatekeeper.sh/v1beta1
+kind: K8sRestrictRoleBindings
+metadata: # kpt-merge: /restrict-clusteradmin-rolebindings
+  name: restrict-clusteradmin-rolebindings
+  annotations:
+    # This constraint is not certified by CIS.
+    description: "Restricts use of the cluster-admin role."
+    bundles.validator.forsetisecurity.org/cis-k8s-v1.5.1: 5.1.1
+spec:
+  enforcementAction: dryrun # kpt-set: ${enforcementAction}
```
[the remark this answers is not in the document]
Comment (g-awmalik) [reply examples/set-enforcement-action-simple/README.md:40]: Updated

File: examples/set-enforcement-action-simple/README.md [hunk 15]
```diff
@@ -0,0 +1,64 @@
...
+apiVersion: constraints.gatekeeper.sh/v1beta1
+kind: K8sProhibitRoleWildcardAccess
+metadata: # kpt-merge: /prohibit-role-wildcard-access
+  name: prohibit-role-wildcard-access
+  annotations:
+    # This constraint is not certified by CIS.
+    description: "Restricts use of wildcards in Roles and ClusterRoles."
+    bundles.validator.forsetisecurity.org/cis-k8s-v1.5.1: 5.1.3
+spec:
+  enforcementAction: dryrun # kpt-set: ${enforcementAction}
```
[the remark this answers is not in the document]
Comment (g-awmalik) [reply examples/set-enforcement-action-simple/README.md:63]: Updated

File: examples/set-enforcement-action-simple/README.md [hunk 16]
```diff
@@ -0,0 +1,64 @@
...
+```shell
+$ kpt fn render set-enforcement-action-simple
+```
+
+### Expected result
+
+The two policy constraints should have been muted with their `spec.enforcementAction` 
+elements set to `dryrun` which were initially set to `deny`
+
+`restrict-clusteradmin-rolebindings.yaml`
```
[the remark this answers is not in the document]
Comment (g-awmalik) [reply examples/set-enforcement-action-simple/README.md:29]: I was referring to the resource name here. I took it out since it's unclear.

File: examples/set-enforcement-action-simple/README.md [hunk 17]
```diff
@@ -0,0 +1,64 @@
...
+      apiGroup: "rbac.authorization.k8s.io"
+      kind: "ClusterRole"
+      name: "cluster-admin"
+    allowedSubjects:
+    - apiGroup: "rbac.authorization.k8s.io"
+      kind: "Group"
+      name: "system:masters"
+```
+
+`prohibit-role-wildcard-access`
```
[the remark this answers is not in the document]
Comment (g-awmalik) [reply examples/set-enforcement-action-simple/README.md:52]: Same as the above comment.

File: functions/go/set-enforcement-action/README.md (hunk 9 above)
Comment (g-awmalik) [reply functions/go/set-enforcement-action/README.md:12]: Done.

File: functions/go/set-enforcement-action/README.md (hunk 10 above)
Comment (g-awmalik) [reply functions/go/set-enforcement-action/README.md:11]: Done.

File: functions/go/set-enforcement-action/README.md (hunk 11 above)
Comment (g-awmalik) [reply functions/go/set-enforcement-action/README.md:38]: Fixed.

File: functions/go/set-enforcement-action/README.md (hunk 12 above)
Comment (g-awmalik) [reply functions/go/set-enforcement-action/README.md:13]: Reworded. Made bullets instead.

File: functions/go/set-enforcement-action/main.go (hunk 13 above)
Comment (g-awmalik) [reply functions/go/set-enforcement-action/main.go:114]: Sorry, my bad. Added constant for `warn` now.

**Comment by g-awmalik** [comment]
@mengqiy - 3rd time's a charm!

**Review by mengqiy — APPROVED** [review]
LGTM
